### ISSUE TITLE
Password derivation functions

### ISSUE THREAD

**Issue by rvangraan** [potatosalad/erlang-libsodium#1]
Hi there,

Very happy to have discovered this implementation. Good work! I managed to get it compiled and workin on my Mac on the first try.

I am interested in using this library for a project, but I need the password functions:
https://download.libsodium.org/doc/password_hashing/index.html

I am not really qualified to implement the C code, so was wondering how much effort it would be to add the key derivation functions? 
```
int crypto_pwhash_scryptsalsa208sha256(unsigned char * const out,
                                       unsigned long long outlen,
                                       const char * const passwd,
                                       unsigned long long passwdlen,
                                       const unsigned char * const salt,
                                       unsigned long long opslimit,
                                       size_t memlimit);
```
and perhaps:
```
int crypto_pwhash_scryptsalsa208sha256_str(char out[crypto_pwhash_scryptsalsa208sha256_STRBYTES],
                                           const char * const passwd,
                                           unsigned long long passwdlen,
                                           unsigned long long opslimit,
                                           size_t memlimit);
``` 
and
```
int crypto_pwhash_scryptsalsa208sha256_str_verify(const char str[crypto_pwhash_scryptsalsa208sha256_STRBYTES],
                                                  const char * const passwd,
                                                  unsigned long long passwdlen);
```
Rudolph

**Comment by potatosalad** [comment]
@rvangraan I just published [version 0.0.4 to hex.pm](https://hex.pm/packages/libsodium/0.0.4) with support for `crypto_pwhash_scryptsalsa208sha256` (see commit fb93f36fa9543ad6f541a02066370a59b6c258e1).

Let me know if this works for you and we'll close this issue.

#### Usage

##### `libsodium_crypto_pwhash_scryptsalsa208sha256:crypto_pwhash_scryptsalsa208sha256/5`

Test variables are from [`test/default/pwhash.c` from libsodium](https://github.com/jedisct1/libsodium/blob/1.0.8/test/default/pwhash.c#L19-L25).

```erlang
% Test Variables
Outlen = 155,
Passwd = libsodium_utils:hex2bin(<<"a347ae92bce9f80f6f595a4480fc9c2fe7e7d7148d371e9487d75f5c23008ffae065577a928febd9b1973a5a95073acdbeb6a030cfc0d79caa2dc5cd011cef02c08da232d76d52dfbca38ca8dcbd665b17d1665f7cf5fe59772ec909733b24de97d6f58d220b20c60d7c07ec1fd93c52c31020300c6c1facd77937a597c7a6">>),
Salt = libsodium_utils:hex2bin(<<"5541fbc995d5c197ba290346d2c559dedf405cf97e5f95482143202f9e74f5c2">>),
Opslimit = 481326,
Memlimit = 7256678.

% Expected Output
Out = libsodium_utils:hex2bin(<<"8d40f5f8c6a1791204f03e19a98cd74f918b6e331b39cfc2415e5014d7738b7bb0a83551fb14a035e07fdd4dc0c60c1a6822ac253918979f6324ff0c87cba75d3b91f88f41ca5414a0f152bdc4d636f42ab2250afd058c19ec31a3374d1bd7133289bf21513ff67cbf8482e626aee9864c58fd05f9ea02e508a10182b7d838157119866f072004987ef6c56683ed207705923921af9d76444a331a">>).

% Should be true
Out =:= libsodium_crypto_pwhash_scryptsalsa208sha256:crypto_pwhash_scryptsalsa208sha256(Outlen, Passwd, Salt, Opslimit, Memlimit).
```

##### `libsodium_crypto_pwhash_scryptsalsa208sha256:str/3` and `libsodium_crypto_pwhash_scryptsalsa208sha256:str_verify/2`

```erlang
% Test Variables
Passwd = <<"Correct Horse Battery Staple">>,
Opslimit = 481326,
Memlimit = 7256678.

% Create Password Hash
Str = libsodium_crypto_pwhash_scryptsalsa208sha256:str(Passwd, Opslimit, Memlimit).

% Verify Password against Hash (-1 means false)
0 =:= libsodium_crypto_pwhash_scryptsalsa208sha256:str_verify(Str, Passwd).
```

##### `libsodium_crypto_pwhash_scryptsalsa208sha256:ll/6`

Low-level password hashing.  Produces the same output as the first example.

```erlang
% Test Variables
Passwd = libsodium_utils:hex2bin(<<"a347ae92bce9f80f6f595a4480fc9c2fe7e7d7148d371e9487d75f5c23008ffae065577a928febd9b1973a5a95073acdbeb6a030cfc0d79caa2dc5cd011cef02c08da232d76d52dfbca38ca8dcbd665b17d1665f7cf5fe59772ec909733b24de97d6f58d220b20c60d7c07ec1fd93c52c31020300c6c1facd77937a597c7a6">>),
Salt = libsodium_utils:hex2bin(<<"5541fbc995d5c197ba290346d2c559dedf405cf97e5f95482143202f9e74f5c2">>),
N = 4096,
R = 8,
P = 3,
Buflen = 155.

% Expected Output
Buf = libsodium_utils:hex2bin(<<"8d40f5f8c6a1791204f03e19a98cd74f918b6e331b39cfc2415e5014d7738b7bb0a83551fb14a035e07fdd4dc0c60c1a6822ac253918979f6324ff0c87cba75d3b91f88f41ca5414a0f152bdc4d636f42ab2250afd058c19ec31a3374d1bd7133289bf21513ff67cbf8482e626aee9864c58fd05f9ea02e508a10182b7d838157119866f072004987ef6c56683ed207705923921af9d76444a331a">>).

% Should be true
Buf =:= libsodium_crypto_pwhash_scryptsalsa208sha256:ll(Passwd, Salt, N, R, P, Buflen).
```

**Comment by rvangraan** [comment]
Wow thank you very much! I've just tried it and I got the same results. Really appreciates the effort!

Thanks, Rudolph
